libraries/AP_HAL_ChibiOS/UARTDriver.cpp

```diff
@@ -948,19 +941,32 @@ void UARTDriver::write_pending_bytes_DMA(uint32_t n)
 
             const uint32_t tx_size = dmaStreamGetTransactionSize(txdma);
 
-            if (tx_size == 0) {
-                // There are no bytes left so we are done
-                _last_write_completed_us = AP_HAL::micros();
+            if (tx_size >= tx_len) {
+                // we didn't write any of our bytes
+                tx_len = 0;
             } else {
-                // an actual timeout occurred, record how much was sent
-                // tx_size is how much was not sent (could be 0)
-                tx_len = MIN(tx_len, tx_size);
+                // record how much was sent tx_size is how much was
+                // not sent (could be 0)
+                tx_len -= tx_size;
```

Comment:
I don't think this is correct, this bug was in earlier versions of the code. tx_size is how much could *not* be sent rather than how much was sent.

Reply:
right .... to we wanted to send tx_len, we didn't send tx_size, so the amount we actually sent is tx_len - tx_size
I'm pretty sure this is right